orttraining/orttraining/models/runner/pipeline.h

```diff
@@ -16,78 +16,162 @@ namespace onnxruntime {
+  void AddCompute(const int batch_id, const Action::Pass pass, const int upstream_time = -1, const int upstream_stage = -1);
+
+  bool IsEmpty() const;
+  size_t NumActions() const;
+  bool HasCompute() const;
+  bool HasRendTo(const int stage) const;
+  bool HasRecvFrom(const int stage) const;
+  Action& operator[](int index);
+  const Action& operator[](int index) const;
+  Action& GetFrontAction();
+  const Action& GetFrontAction() const;
+
+  // Print this structure following a fixed-length format.
+  // It assumes there are at most 2 actions per slot.
+  friend std::ostream& operator<<(std::ostream& stream, const PipelineSlot& slot);
+  void SetWaitedEvent(const std::vector<int> event);
+  std::vector<int> GetWaitedEvent() const;
+  void SetRecordedEvent(const std::vector<int> event);
+  std::vector<int> GetRecordedEvent() const;
 
  private:
-  std::vector<int> SearchLastRecordedEvents(int time_id, int stage_id) const;
+  // Actions which can be executed in parallel in this time slot.
+  std::vector<Action> operators_;
+
+  // For MPI PipeDream schedule, it's used to support Wait -> Recv -> Wait -> Compute -> Record -> Send -> Record.
+  // Since Send, Recv, and Compute are stored in the same slot, each slot contains two waited events and two recorded events.
+  //
+  // For NCCL PipDream schedule, it's used to support Wait -> Recv -> Record -> Wait -> Compute -> Record -> Wait -> Send -> Record.
+
+  // Events waited by this slot. 
+  std::vector<int> waited_events_;
+  // Events recorded by this slot. 
+  std::vector<int> recorded_events_;
+};
+
+class PipelineScheduler {
+ public:
+  PipelineScheduler(int num_batches, int num_stages);
+  size_t GetScheduleSize() const;
+  size_t GetStageSize() const;
+  // APIs to get NCCL event for
+  // Wait -> Recv -> Record -> Wait -> Compute -> Record -> Wait -> Send -> Record.
+  int GetForwardComputeWaitedEvent(const int batch_id, const int stage_id) const;
+  int GetForwardComputeRecordedEvent(const int batch_id, const int stage_id) const;
+  int GetBackwardComputeWaitedEvent(const int batch_id, const int stage_id) const;
+  int GetBackwardComputeRecordedEvent(const int batch_id, const int stage_id) const;
+  int GetForwardSendWaitedEvent(const int batch_id, const int stage_id) const;
+  int GetForwardSendRecordedEvent(const int batch_id, const int stage_id) const;
+  int GetBackwardSendWaitedEvent(const int batch_id, const int stage_id) const;
+  int GetBackwardSendRecordedEvent(const int batch_id, const int stage_id) const;
+  int GetForwardRecvWaitedEvent(const int batch_id, const int stage_id) const;
+  int GetForwardRecvRecordedEvent(const int batch_id, const int stage_id) const;
+  int GetBackwardRecvWaitedEvent(const int batch_id, const int stage_id) const;
+  int GetBackwardRecvRecordedEvent(const int batch_id, const int stage_id) const;
```

Comment:
would it better to use one function to return a struct at given batch_id and stage_id?

Reply:
Having MpiPipelineScheduler and NcclPipelineScheduler is doable but I'd like to first have NCCL P2P working and then come back to refactorize these classes again. Would this be ok to you?

---
In reply to: [455384433](https://github.com/microsoft/onnxruntime/pull/4524#discussion_r455384433) [](ancestors = 455384433)

Reply:
Sure.

---
In reply to: [455584382](https://github.com/microsoft/onnxruntime/pull/4524#discussion_r455584382) [](ancestors = 455584382,455384433)